### PR TITLE
refactoring: SA Maven Utility Functions | Utils Arch changes

### PR DESCRIPTION
**Pull request by deepak1725**
Optimization SA Utility Functions

Jira: https://issues.redhat.com/browse/APPAI-1527

### CODE REVIEW

**Comment by deepak1725** [comment]
[test]

File: f8a_utils/tree_generator.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+                "manifest_file_path": manifest['filepath'],
+                "manifest_file": manifest['filename']
+            }
+            resolved = []
+            data = manifest['content']
+
+            if isinstance(data, bytes):
+                data = data.decode("utf-8")
+
+            tree = self._get_tree(data)
```
Comment (prashbnair) [inline f8a_utils/tree_generator.py:43]: rename to get_transitives

File: f8a_utils/tree_generator.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+                resolved.append(tmp_json)
+            dep['_resolved'] = resolved
+            details.append(dep)
+            details_json = {"details": details}
+            result.append(details_json)
+
+        deps['result'] = result
+        return deps
+
+    def _get_transitives(self, transitives: list) -> list:
```
Comment (prashbnair) [inline f8a_utils/tree_generator.py:64]: parse_transitives

File: f8a_utils/tree_generator.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            resolved = []
+            data = manifest['content']
+
+            if isinstance(data, bytes):
+                data = data.decode("utf-8")
+
+            tree = self._get_tree(data)
+            for direct, transitives in tree.items():
+                # Add meta data to generated tree.
+                parsed_json = self._parse_string(direct)
```
Comment (prashbnair) [inline f8a_utils/tree_generator.py:46]: Is the conditon for 'test' missing?

File: f8a_utils/tree_generator.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+            # groupId:artifactId:packaging:classifier:version:scope
+            a['groupId'], a['artifactId'], a['packaging'], a['classifier'], a['version'], \
+                a['scope'] = coordinates_str.split(':')
+        else:
+            raise ValueError('Invalid Maven coordinates %s', coordinates_str)
+
+        return a
+
+
+class NpmDependencyTreeGenerator(DependencyTreeGenerator):
```
Comment (prashbnair) [inline f8a_utils/tree_generator.py:138]: Npm structure also seems similar to maven and the above code can probably be reused.

**Review by prashbnair — CHANGES_REQUESTED** [review]
I have not reviewed goLang as I do not know the file structure.

File: f8a_utils/tree_generator.py (hunk 1 above)
Comment (deepak1725) [reply f8a_utils/tree_generator.py:43]: it's not really transitives, If you remember we are generating a Parent-Child Relationship tree here and later on iterating over Child (Transitives). 

This method actually generating a Tree here.

File: f8a_utils/tree_generator.py (hunk 3 above)
Comment (deepak1725) [reply f8a_utils/tree_generator.py:46]: Do we really need "Test" condition..? Dont really know why it was put in place first time.. 
For me, it looks like Technical Debt.. https://github.com/fabric8-analytics/fabric8-analytics-utils/blob/master/f8a_utils/dependency_finder.py#L51

File: f8a_utils/tree_generator.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        :param content: file contents from dependency.txt
+        :return: Tree in format ({d1:[t1, t2]})
+        """
+        final_map = {}
+        intermediate_map = defaultdict(list)
+        module = ''
+        for line in content.split("\n"):
+            if '->' in line:
+                prefix, suffix = line.split('->')
+                prefix = prefix.replace('"', '').replace(';', '').strip()
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply f8a_utils/tree_generator.py]: Ack!

File: f8a_utils/dependency_finder.py [hunk 6]
```diff
@@ -10,290 +29,13 @@ class DependencyFinder():
...
-        """Clean Golang Dep."""
-        if isinstance(dependencies, bytes):
-            dependencies = dependencies.decode("utf-8")
-        dependencies = dependencies[:dependencies.rfind('\n')]
-        if not dependencies:
-            raise ValueError('Dependency list cannot be empty')
-        return dependencies.split('\n')
+        if type(show_transitive) is not bool:
+            show_transitive = show_transitive == "true"
+        get_dependency_tree = get_tree(ecosystem)()
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply f8a_utils/dependency_finder.py]: Ack!

File: f8a_utils/tree_generator.py (hunk 2 above)
Comment (deepak1725) [reply f8a_utils/tree_generator.py]: Ack!

File: f8a_utils/tree_generator.py (hunk 1 above)
Comment (prashbnair) [reply f8a_utils/tree_generator.py:43]: Can you then rename to get_dependency_tree?

File: f8a_utils/tree_generator.py (hunk 3 above)
Comment (prashbnair) [reply f8a_utils/tree_generator.py:46]: Can you please check with @dgpatelgit . He had done some refactoring in this area.

File: f8a_utils/tree_generator.py (hunk 1 above)
Comment (deepak1725) [reply f8a_utils/tree_generator.py:43]: Okay, Let me try to explain in a pictorial way.
Blank diagram

File: f8a_utils/dependency_finder.py [hunk 7]
```diff
@@ -10,290 +29,13 @@ class DependencyFinder():
...
-        """Clean Golang Dep."""
-        if isinstance(dependencies, bytes):
-            dependencies = dependencies.decode("utf-8")
-        dependencies = dependencies[:dependencies.rfind('\n')]
-        if not dependencies:
-            raise ValueError('Dependency list cannot be empty')
-        return dependencies.split('\n')
+        if type(show_transitive) is not bool:
+            show_transitive = show_transitive == "true"
+        dependency_tree = get_tree(ecosystem)()
```
Comment (prashbnair) [inline f8a_utils/dependency_finder.py:34]:
```suggestion
        dependency_tree_generator = get_dependency_tree_generator(ecosystem)()
```

File: f8a_utils/tree_generator.py (hunk 1 above)
Comment (prashbnair) [reply f8a_utils/tree_generator.py:43]: I see that you are calling get_dependencies, there is no get_dependency_tree

**Review by prashbnair — CHANGES_REQUESTED** [review]

File: f8a_utils/tree_generator.py (hunk 3 above)
Comment (deepak1725) [reply f8a_utils/tree_generator.py:46]: After Sync up with DGP, Just got to know: `test scope` will be there if tree is generated using 
```
mvn org.apache.maven.plugins:maven-dependency-plugin:3.0.2:tree -DoutputFile=/someloc/dependencies.txt -DoutputType=dot -DappendOutput=true;
``` 
We have to explicitly avoid test modules. Re-Added condition for the same.

File: f8a_utils/tree_generator.py (hunk 1 above)
Comment (deepak1725) [reply f8a_utils/tree_generator.py]: Ack.

File: f8a_utils/dependency_finder.py (hunk 7 above)
Comment (deepak1725) [reply f8a_utils/dependency_finder.py]: ACK.

**Review by prashbnair — APPROVED** [review]

**Pull request merged**